### PR TITLE
ContainerInfraV1: add Cluster template resource

### PR DESCRIPTION
**Pull request by ozerovandrei**
Add resource to manage OpenStack Magnum cluster templates.
Add common resourceLinks() function to unmarhal slice of gophercloud.Links to a slice of maps.

For #402

Some links:
[Constraints for the `apiserver_port`](https://github.com/openstack/magnum/blob/stable/queens/magnum/api/controllers/v1/cluster_template.py#L82)
[Requirement of the `coe`](https://github.com/openstack/magnum/blob/stable/queens/magnum/api/controllers/v1/cluster_template.py#L50)
[Requirement of the `image`](https://github.com/openstack/magnum/blob/stable/queens/magnum/api/controllers/v1/cluster_template.py#L53)

### CODE REVIEW

**Comment by ozerovandrei** [comment]
It's almost ready, I just want to resolve to more issues:

1) Right now the cluster template resource gets the `labels` attribute as a string of key-value attributes (as we can do via `openstack coe cluster template create --labels kube_tag =1.11.1, prometheus_monitoring =true,...`). I think that this will be uncomfortable for users to see such long lines in their hcl manifests.
I want to make an input as a map:
```hcl
labels = {
    kube_tag = "1.11.1"
    prometheus_monitoring = "true"
    influx_grafana_dashboard_enabled = "true"
    kube_dashboard_enabled = "true"
}
```
And then convert it to the string inside a provider resource.

2) Second issue I can't fix right now and I need some help.
So, Magnum API gets UpdateOpts as a slice of add/replace/delete values: [doc.go#L71](https://github.com/gophercloud/gophercloud/blob/master/openstack/containerinfra/v1/clustertemplates/doc.go#L71).
I tried to use the replace action on every attribute that is changed (checking that with the usual `d.HasChange()`) and use the delete action if new value of the attribute is an empty string.
But for some reason `d.HasChange` ignores new empty value. (So i just use the replace action for now).
I will do some digging about it but maybe someone can point me in the right direction? @jtopjian do you know some workarounds for it?

**Comment by jtopjian** [comment]
> But for some reason d.HasChange ignores new empty value.

It shouldn't. `HasChange` should notify whenever any sort of change to a previously set argument happens.

Can you provide an example that reproduces this behavior?

**Comment by ozerovandrei** [comment]
For example, we created the following cluster template:
```hcl
resource "openstack_containerinfra_clustertemplate_v1" "clustertemplate_1" {
  name = "clustertemplate_1"
  image = "Fedora-Atomic-27"
  coe = "kubernetes"
  http_proxy = "localhost:8800"
}
```

And after that we want to set the `http_proxy` attribute to `""`.

**Comment by jtopjian** [comment]
The reason is because `http_proxy` is set to `Computed`. This example is timely because I've recently begun questioning and being strict in PRs about whether an argument really needs to be Computed or not. There have been too many bugs introduced because of this.

Unless the Magnum API sets a default proxy, it does not need to be computed.

The reason `HasChange` isn't triggering a change is because although you have removed the value, Terraform thinks Magnum is going to send a computed value.

Try this out by making `http_proxy` not computed.

A better example might be `project_id` and `user_id`. Let's assume an admin user can set both of these to assign a template to a different project. So an admin would run Terraform with these values set:

```hcl
resource "openstack_containerinfra_clustertemplate_v1" "clustertemplate_1" {
  name = "clustertemplate_1"
  image = "Fedora-Atomic-27"
  coe = "kubernetes"
  http_proxy = "localhost:8800"

  project_id = "some-other-project"
  user_id = "some-other-user"
}
```

But if these arguments are not set, then the API will return the project and user who created the template. 

A good rule of thumb for determining if an argument should be computed is: who is authoritative of the data: the user or API service? If a user can set, change, and remove a value, then _usually_ (there are, of course, some exceptions) the argument should not be computed. If a user can set and change a value, but the API can set a default value, then computed should be fine. If the user _cannot_ set a value and only the API sets it, then it is a "read-only" argument (see the data sources for examples of this).

Make sense? If so, you can see how misuse of "computed" can lead to bugs. :(

One common misconception about "computed" arguments is that you cannot use `d.Set()` unless the value is computed. This isn't true. I think this dates back to before Terraform had "import" support. Back then, you would only run `d.Set` on computed fields because the user defined everything else, so why do `d.Set` again? But with importing, you are emulating an existing tf configuration, so you must set non-computed fields, too.

Let me know if you have any questions.

**Comment by ozerovandrei** [comment]
@jtopjian this is awesome explanation, thank you!

**Comment by ozerovandrei** [comment]
This is ready.
Although, we need the `OpenStack Magnum` installed in the OpenLab acceptance tests runner.

**Comment by jtopjian** [comment]
@ozerovandrei Cool! I hope to be able to review this in the next few days. My schedule has been really busy lately.

File: openstack/resource_openstack_containerinfra_clustertemplate_v1.go [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+	d.Set("registry_enabled", s.RegistryEnabled)
+	d.Set("server_type", s.ServerType)
+	d.Set("tls_disabled", s.TLSDisabled)
+	d.Set("volume_driver", s.VolumeDriver)
+	d.Set("region", GetRegion(d, config))
+	d.Set("name", s.Name)
+	d.Set("project_id", s.ProjectID)
+	d.Set("user_id", s.UserID)
+	d.Set("created_at", s.CreatedAt)
+	d.Set("updated_at", s.UpdatedAt)
```
Comment (jtopjian) [inline openstack/resource_openstack_containerinfra_clustertemplate_v1.go]: For both of these, you'll need to format them into a string. If you do:

```go
if err := d.Set("created_at", s.CreatedAt); err != nil {
  log.Printf("[DEBUG] created_at: %s", err)
}
```

You'll see an error message in the logs. In addition, you'll notice that these are not being set in state, either.

Alternatively, we can just remove these fields unless you see a need to have them.

**Comment by jtopjian** [comment]
@ozerovandrei also, with regard to the test failure, we'll want to add something similar to this:

https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/provider_test.go#L114

Then when OpenLab adds Magnum support, they can set the environment variable on their side.

**Comment by ozerovandrei** [comment]
@jtopjian Thank you for the review.

I've tried the [IntBetween](https://github.com/hashicorp/terraform/blob/v0.11.7/helper/validation/validation.go#L18) validation function and got the following problem:
 - we need to vendor `github.com/hashicorp/terraform/helper/validation`
 - if we add it to vendor, terraform-provider-openstack will be broken, it will throw panic:
```sh
panic: gob: registering duplicate types for "github.com/zclconf/go-cty/cty.primitiveType": cty.primitiveType != cty.primitiveType

goroutine 1 [running]:
encoding/gob.RegisterName(0xc42003c7b0, 0x2b, 0x12c7d20, 0xc420290937)
	/usr/lib/go-1.10/src/encoding/gob/type.go:820 +0x5d4
github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty.init.3()
	/opt/stack/go/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/types_to_register.go:52 +0x5bf
exit status 2
FAIL	github.com/terraform-providers/terraform-provider-openstack/openstack	0.010s
```
 - we can fix this, by fetching the latest full `github.com/hashicorp/terraform` with all of its dependencies.

Can I do such big vendoring fetching inside this PR?

**Comment by jtopjian** [comment]
Did you do the vendoring with a v0.11.7 tag?

If so, yuck. You're right - we definitely don't want to do this large of an update in this PR. We're going to have to do a vendor update when Terraform 0.12 comes out, so let's hold off until then.

Thank you for looking into this.

**Comment by ozerovandrei** [comment]
Of course not 😄 
Yeah, it's working with the `v0.11.7` tag, all good.

**Comment by ozerovandrei** [comment]
This can be reviewed again.
Hope that I resolved all issues.

File: openstack/resource_openstack_containerinfra_clustertemplate_v1.go [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+	d.Set("name", s.Name)
+	d.Set("project_id", s.ProjectID)
+	d.Set("user_id", s.UserID)
+
+	if err := d.Set("created_at", s.CreatedAt); err != nil {
+		log.Printf("[DEBUG] created_at: %s", err)
+	}
+	if err := d.Set("updated_at", s.CreatedAt); err != nil {
+		log.Printf("[DEBUG] updated_at: %s", err)
+	}
```
Comment (jtopjian) [inline openstack/resource_openstack_containerinfra_clustertemplate_v1.go]: If you look in the debug logs, you'll see something similar to:

```
2018/08/30 15:57:42 [DEBUG] created_at: created_at: '' expected type 'string', got unconvertible type 'time.Time'
```

So we need to do:

```
d.Set("created_at", s.CreatedAt.Format(time.RFC3339))
```

**Comment by ozerovandrei** [comment]
recheck

**Comment by jtopjian** [comment]
@ozerovandrei LGTM!

I want to get #409 merged first since it's a bug fix and has priority. Once that's merged, feel free to merge this one when you're ready :)

**Comment by ozerovandrei** [comment]
@jtopjian got it.

**Comment by jtopjian** [comment]
@ozerovandrei looks like there's a vendor conflict - I figured that might happen.

Once it's resolved, feel free to merge whenever you're ready.

After it's merged, you can edit the CHANGELOG file directly through Github to add the release note. Let me know if you have any questions.

**Pull request merged**